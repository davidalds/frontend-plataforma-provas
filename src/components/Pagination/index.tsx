import { Button, ButtonGroup, HStack, Icon, IconButton } from '@chakra-ui/react'
import { useEffect, useState } from 'react'
import { AiOutlineArrowLeft, AiOutlineArrowRight } from 'react-icons/ai'
import { IPropsPagination } from './interfaces/pagination'

const Pagination = ({
  totalDataElements,
  limitPerPage,
  totalShowedButtonsInPagination,
  offsetFc,
}: IPropsPagination) => {
  const totalElements = totalDataElements
  const limit = limitPerPage
  const pagesTotal = Number.isInteger(totalElements / limit)
    ? totalElements / limit
    : Math.ceil(totalElements / limit)
  const buttonsLimit =
    totalShowedButtonsInPagination > pagesTotal
      ? pagesTotal
      : totalShowedButtonsInPagination
  const [page, setPage] = useState<number>(1)
  const [indexSum, setIndexSum] = useState<number>(1)
  const [minButtonValue, setMinButtonValue] = useState<number>(1)
  const [maxButtonValue, setMaxButtonValue] = useState<number>(buttonsLimit)

  useEffect(() => {
    offsetFc((page - 1) * limit)
    if (page < minButtonValue) {
      handleChangeStatesButtons(-1)
    }

    if (page > maxButtonValue) {
      handleChangeStatesButtons(1)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [page])

  const handleChangeStatesButtons = (n: number) => {
    setIndexSum((prevState) => prevState + n)
    setMinButtonValue((prevState) => prevState + n)
    setMaxButtonValue((prevState) => prevState + n)
  }

  const handleCustomPage = (n_value: number) => {
    setPage(n_value)
  }

  const handlePrevPage = () => {
    if (page > 1) {
      setPage((prevState) => prevState - 1)
    }
  }

  const handleNextPage = () => {
    if (page < pagesTotal) {
      setPage((prevState) => prevState + 1)
    }
  }

  const handleStart = () => {
    if (page > 1) {
      setPage(1)
      setMinButtonValue(1)
      setMaxButtonValue(buttonsLimit)
      setIndexSum(1)
    }
  }

  const handleEnd = () => {
    if (page !== pagesTotal) {
      setPage(pagesTotal)
      setMinButtonValue(pagesTotal - (buttonsLimit - 1))
      setMaxButtonValue(pagesTotal)
      setIndexSum(pagesTotal - (buttonsLimit - 1))
    }
  }

  return (
    <>
      <HStack p={2} justifyContent={'center'} m={4}>
        <Button
          size={'sm'}
          colorScheme={'blue'}
          variant={'outline'}
          isDisabled={page === 1}
          onClick={handleStart}
        >
          Início
        </Button>
        <IconButton
          aria-label="anterior"
          icon={<Icon as={AiOutlineArrowLeft} />}
          colorScheme={'blue'}
          size={'sm'}
          isDisabled={page === 1}
          onClick={handlePrevPage}
        />
        <ButtonGroup spacing={2}>
          {minButtonValue !== 1 ? <span>...</span> : <></>}
          {Array.from({ length: buttonsLimit }).map((_, index) => (
            <Button
              key={index + indexSum}
              variant={'outline'}
              colorScheme={'blue'}
              size={'sm'}
              rounded={'50%'}
              onClick={() => handleCustomPage(index + indexSum)}
              isActive={page === index + indexSum}
            >
              {index + indexSum}
            </Button>
          ))}
          {maxButtonValue < pagesTotal ? <span>...</span> : <></>}
        </ButtonGroup>
        <IconButton
          aria-label="próximo"
          icon={<Icon as={AiOutlineArrowRight} />}
          colorScheme={'blue'}
          size={'sm'}
          isDisabled={page === Math.ceil(totalElements / limit)}
          onClick={handleNextPage}
        />
        <Button
          size={'sm'}
          colorScheme={'blue'}
          variant={'outline'}
          isDisabled={page === Math.ceil(totalElements / limit)}
          onClick={handleEnd}
        >
          Final
        </Button>
      </HStack>
    </>
  )
}

export default Pagination
