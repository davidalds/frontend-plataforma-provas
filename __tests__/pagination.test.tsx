import { screen, render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'
import Pagination from 'components/Pagination'

describe('<Pagination />', () => {
  beforeAll(() => {
    expect(
      <Pagination
        totalDataElements={100}
        limitPerPage={10}
        totalShowedButtonsInPagination={5}
        offsetFc={jest.fn()}
      />
    ).toMatchSnapshot()
  })

  test('test showed buttons', async () => {
    const { debug } = render(
      <Pagination
        totalDataElements={100}
        limitPerPage={10}
        totalShowedButtonsInPagination={5}
        offsetFc={jest.fn()}
      />
    )

    const buttons = screen.getAllByRole('button', { name: /\d/ })

    expect(buttons).toHaveLength(5)
  })

  test('test disabled buttons', async () => {
    const { debug } = render(
      <Pagination
        totalDataElements={10}
        limitPerPage={10}
        totalShowedButtonsInPagination={1}
        offsetFc={jest.fn()}
      />
    )

    const beginBtn = screen.getByRole('button', { name: /Início/i })
    const endBtn = screen.getByRole('button', { name: /Final/i })
    const prevBtn = screen.getByLabelText('anterior')
    const nextBtn = screen.getByLabelText('próximo')

    expect(beginBtn).toBeDisabled()
    expect(endBtn).toBeDisabled()
    expect(prevBtn).toBeDisabled()
    expect(nextBtn).toBeDisabled()
  })

  test('test ellipsis', async () => {
    const { debug } = render(
      <Pagination
        totalDataElements={30}
        limitPerPage={10}
        totalShowedButtonsInPagination={1}
        offsetFc={jest.fn()}
      />
    )

    expect(screen.getAllByText('...')).toHaveLength(1)

    const nextBtn = screen.getByLabelText('próximo')

    await userEvent.click(nextBtn)

    expect(screen.getAllByText('...')).toHaveLength(2)
  })

  test('test begin and end button', async () => {
    const { debug } = render(
      <Pagination
        totalDataElements={100}
        limitPerPage={10}
        totalShowedButtonsInPagination={5}
        offsetFc={jest.fn()}
      />
    )

    const beginBtn = screen.getByRole('button', { name: /Início/i })
    const endBtn = screen.getByRole('button', { name: /Final/i })

    expect(
      screen.getAllByRole('button', { name: /\d/ }).pop()
    ).toHaveTextContent('5')

    await userEvent.click(endBtn)

    expect(
      screen.getAllByRole('button', { name: /\d/ }).pop()
    ).toHaveTextContent('10')

    await userEvent.click(beginBtn)

    expect(
      screen.getAllByRole('button', { name: /\d/ }).shift()
    ).toHaveTextContent('1')
  })

  test('test number of buttons bigger than available pages', async () => {
    const { debug } = render(
      <Pagination
        totalDataElements={100}
        limitPerPage={10}
        totalShowedButtonsInPagination={12}
        offsetFc={jest.fn()}
      />
    )

    expect(screen.getByRole('group').children).toHaveLength(10)
  })
})
