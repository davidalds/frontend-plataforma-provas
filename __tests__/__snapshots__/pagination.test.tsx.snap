// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<Pagination
  limitPerPage={10}
  offsetFc={[MockFunction]}
  totalDataElements={100}
  totalShowedButtonsInPagination={5}
/>
`;
